python/mead/downloader.py

```diff
@@ -0,0 +1,310 @@
+            return False
+    return True
+
+
+def read_json(filepath):
+    if not os.path.exists(filepath):
+        return {}
+    with open(filepath) as f:
+        j_con = json.load(f)
+    return j_con
+
+
+def write_json(content, filepath):
+    with open(filepath, "w") as f:
+        json.dump(content, f, indent=True)
+
+
+class Downloader(object):
+    def __init__(self, data_download_cache, cache_ignore):
+        super(Downloader, self).__init__()
+        self.cache_ignore = cache_ignore
+        self.data_download_cache = data_download_cache
+
+    def download(self):
+        pass
+
+
+class SingleFileDownloader(Downloader):
+    def __init__(self, dataset_file, data_download_cache, cache_ignore=False):
+        super(SingleFileDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_file = dataset_file
+        self.data_download_cache = data_download_cache
+
+    def download(self):
+        file_loc = self.dataset_file
+        if is_file_correct(file_loc):
+            return file_loc
+        elif validate_url(file_loc):  # is it a web URL? check if exists in cache
+            url = file_loc
+            dcache_path = os.path.join(self.data_download_cache, DATA_CACHE_CONF)
+            dcache = read_json(dcache_path)
+            if url in dcache and is_file_correct(dcache[url], self.data_download_cache, url) and not self.cache_ignore:
+                LOG("file for {} found in cache, not downloading".format(url))
+                return dcache[url]
+            else:  # download the file in the cache, update the json
+                cache_dir = self.data_download_cache
+                LOG("using {} as data/embeddings cache".format(cache_dir))
+                temp_file = web_downloader(url)
+                dload_file = extractor(filepath=temp_file, cache_dir=cache_dir,
+                                       extractor_func=zipd.get(mime_type(temp_file), None))
+                dcache.update({url: dload_file})
+                write_json(dcache, os.path.join(self.data_download_cache, DATA_CACHE_CONF))
+                return dload_file
+        else:
+            raise RuntimeError("the file {} is not in cache and can not be downloaded".format(file_loc))
+
+
+class DataDownloader(Downloader):
+    def __init__(self, dataset_desc, data_download_cache, enc_dec=False, cache_ignore=False):
+        super(DataDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.dataset_desc = dataset_desc
+        self.data_download_cache = data_download_cache
+        self.enc_dec = enc_dec
+
+    def download(self):
+        dload_bundle = self.dataset_desc.get("download", None)
+        if dload_bundle is not None:  # download a zip/tar/tar.gz directory, look for train, dev test files inside that.
+            dcache_path = os.path.join(self.data_download_cache, DATA_CACHE_CONF)
+            dcache = read_json(dcache_path)
+            if dload_bundle in dcache and \
```

Comment:
i'm wondering if it would be easier to read this line if `is_dir_correct` was wrapped in an internal method so that the instance variables don't have to be passed to it.